### DISCUSSION TITLE
Clarification around JSVM boot and handler process singleton sharing

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#5773] [Q&A] [answered]
It seems to me that any singleton added to the VM during the boot goroutine isn't available later during the routes.  To have a singleton available throughout it is necessary to store the object in the app store during initialization and then to retrieve that value, cast it to the appropriate type - that is if you want to define a clean interface in Golang code that hides the details of how it's stored.  Otherwise just do all of the app store access within the VM.

**Comment by ganigeorgiev** [comment]
It is not clear what is the problem or what actually you mean by _"singleton added to the VM during the boot goroutine"_. Even "singleton" is too broad of a term to understand what this object is supposed to do.

In general, If you have a JSVM `OnInit` function in your Go code that registers a global variable then it should be accessible from anywhere. The "sharedBinds" are registered in https://github.com/pocketbase/pocketbase/blob/b11ea604363b9c6309c3b0def2277513fa3e3a16/plugins/jsvm/jsvm.go#L276-L285

So for example, if you have:

```go
jsvm.MustRegister(app, jsvm.Config{
    OnInit: func(vm *goja.Runtime) {
        vm.Set("test", 123)
    },
    ...
})
```

And a JS pb_hooks file like:

```js
console.log(test);

routerAdd("GET", "/example", (e) => {
    e.string(200, test)
})
```

This should:
- print "123" to the stdout with the application start
- print "123" as response when accessing the `GET /example`

Whether the access to the `test` variable is concurrent safe depends what actually you are using it for. If it just for reads, then it should be safe. If you have writes you may need a mutex.

**Reply by dschissler** [reply]
This is the way that I'm having a singleton **struct** that can be set on the JS boot that is accessible within a route.

```golang
const BankStoreKey = "pocketcam"

func Set(app core.App, vm *goja.Runtime) {
	obj := vm.NewObject()
	vm.Set("pocketcam", obj)

	bank := pocketcam.NewBank(app)
	app.Store().Set(BankStoreKey, bank)

	obj.Set("new",
		func(device string, opts pocketcam.Options) (*pocketcam.Pocketcam, error) {
			return getBank(app).New(device, opts)
		},
	)

	obj.Set("add",
		func(pc *pocketcam.Pocketcam) error {
			return getBank(app).Add(pc)
		},
	)

	obj.Set("get",
		func(device string, opts pocketcam.Options) (*pocketcam.Pocketcam, error) {
			return getBank(app).Get(device)
		},
	)

	obj.Set("FORMAT_YUYV", pocketcam.FORMAT_YUYV)
	obj.Set("FORMAT_MJPEG", pocketcam.FORMAT_MJPEG)
}

func getBank(app core.App) *pocketcam.Bank {
	return app.Store().Get(BankStoreKey).(*pocketcam.Bank)
}
```

I was doing it like this below, but the cams that were set on the boot JS weren't persisting into the route JS.  So you can see above that I'm putting it all onto the app store.

```golang
func Set(app core.App, vm *goja.Runtime) {
	obj := vm.NewObject()
	vm.Set("pocketcam", obj)

	bank := pocketcam.NewBank(app)

	obj.Set("new", bank.New)
	obj.Set("add", bank.Add)
	obj.Set("get", bank.Get)

	obj.Set("FORMAT_YUYV", pocketcam.FORMAT_YUYV)
	obj.Set("FORMAT_MJPEG", pocketcam.FORMAT_MJPEG)
}
```

Originally I was trying to setup the cameras within a `require` call.  I noticed that it was being called once during boot and once on the first route access.  So I quickly scrapped that to discover the next difficulty.

Note: None of this is really crucial for me.  It's just a hobby project to test boundaries.  I'm currently able to turn a $5 usb camera into a Pocketbase MJPEG  http "video stream".  I'm exclusively setting up routes and the like in the JSVM..

**Reply by ganigeorgiev** [reply]
Please try to be more specific. I know that it takes time to explain things but it is equally  difficult to try to guess what the other person mean. I cannot read your mind and I have literally no context about your project.

What is actually the problem?

What is the desired behavior?

Do you need the dedicate goja objects, aka. can't you use directly the `bank` instance or you want to restrict the access of public methods and fields?

If the first `Set` function is called in the `jsvm.Config.OnInit`, this function will be invoked for every new `goja.Runtime`, meaning that `pocketcam.NewBank(app)` will be called multiple times, aka. there is chance `BankStoreKey` to be overwritten over and over. Is this intended?

If the goal is `pocketcam.NewBank(app)` to be initialized only once then you can try to wrap its creation with [`app.Store().GetOrSet()`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.23.0-rc8/tools/store#Store.GetOrSet) _(available with v0.23.0-rc+)_:

```go
bank := app.Store().GetOrSet(BankStoreKey, func() *pocketcam.Bank {
    // this is expected to be invoked only once (unless BankStoreKey is removed from the app store)
    return pocketcam.NewBank(app)
})
```

**Reply by dschissler** [reply]
That seems like what I'm looking for, but there is a type error: `pkg/pocketcam/jsvm/jsvm.go:16:45: cannot use func() *pocketcam.Bank {…} (value of type func() *pocketcam.Bank) as func() any value in argument to app.Store().GetOrSet`.

The app store seems to be using `any` as the type.  Should I change this to `any` return value and then cast the result afterwards?

**Reply by dschissler** [reply]
To be clear, I'm fine directly accessing the bank struct, but I need there to simply be only a single one for all of the goroutines.  So I'm realizing that not only does this need to be stored on the app store, but that the `GetOrSet` needs to be used in order to make sure that each goroutine isn't simply clobbering the previous store value.  I'm currently just barely hitting the server, which explains why I'm only activating a single JS goroutine on every refresh.

**Reply by ganigeorgiev** [reply, marked as answer]
Yes, I forgot that Go doesn't support type parameters (generics) in struct methods and it uses the generic type from the struct initialization (`map[string]T`), which for the app store is `map[string]any`, or in other words the above example should change to:

```go
bank := app.Store().GetOrSet(BankStoreKey, func() any {
    return pocketcam.NewBank(app)
}).(*pocketcam.Bank)
```

**Reply by dschissler** [reply]
I've  tested this N=2 concurrency.  What I meant about using `bank` directly was like `obj.Set("new", bank.New)` instead of the extra level of wrapping.  It fails at N=3 now, but that is probably a bad message broker design somehow.

```golang
const BankStoreKey = "pocketcam"

func Set(app core.App, vm *goja.Runtime) {
	obj := vm.NewObject()
	vm.Set("pocketcam", obj)

	bank := app.Store().GetOrSet(BankStoreKey, func() any {
		return pocketcam.NewBank(app)
	}).(*pocketcam.Bank)

	obj.Set("new", bank.New)
	obj.Set("get", bank.Get)

	obj.Set("FORMAT_YUYV", pocketcam.FORMAT_YUYV)
	obj.Set("FORMAT_MJPEG", pocketcam.FORMAT_MJPEG)
}
```
